test/gtest/ucs/test_memtype_cache.cc

```diff
@@ -198,6 +198,86 @@ class test_memtype_cache : public ucs::test_with_param<ucs_memory_type_t> {
+                                                      reinterpret_cast<size_t>(iter->end) -
+                                                      interval_end_offset,
+                                                      UCS_MEMORY_TYPE_LAST));
+        }
+
+        // add regions that will be removed as remaining intervals
+        for (std::vector<region_info>::const_iterator iter =
+                 insert_regions.begin(); iter != insert_regions.end(); ++iter) {
+            if (interval_start_offset) {
+                remove_regions.push_back(make_region_info(reinterpret_cast<size_t>(iter->start),
+                                                          reinterpret_cast<size_t>(iter->start) +
+                                                          interval_start_offset,
+                                                          UCS_MEMORY_TYPE_LAST));
+            }
+
+            if (interval_end_offset) {
+                remove_regions.push_back(make_region_info(reinterpret_cast<size_t>(iter->end) -
+                                                          interval_end_offset,
+                                                          reinterpret_cast<size_t>(iter->end),
+                                                          UCS_MEMORY_TYPE_LAST));
+            }
+        }
+    }
+
+    void test_region_insert_and_remove_subintervals(const std::vector<region_info> &regions,
+                                                    size_t interval_start_offset,
+                                                    size_t interval_end_offset,
+                                                    std::vector<region_info> &remove_regions) {
+        generate_test_remove_subintervals(regions, interval_start_offset,
+                                          interval_end_offset, remove_regions);
+
+        // insert new regions
+        for (std::vector<region_info>::const_iterator iter =
+                 regions.begin(); iter != regions.end(); ++iter) {
+            size_t size = UCS_PTR_BYTE_DIFF(iter->start, iter->end);
+            memtype_cache_update(iter->start, size, iter->mem_type);
+            test_ptr_found(iter->start, size, iter->mem_type);
+        }
+
+        // remove subintervals
+        for (std::vector<region_info>::const_iterator iter =
+                 remove_regions.begin(); iter != remove_regions.end(); ++iter) {
+            size_t size = UCS_PTR_BYTE_DIFF(iter->start, iter->end);
+            memtype_cache_remove(iter->start, size);
+            test_ptr_released(iter->start, size);
+        }
+
+        // now all buffers released, check that can't find them
+        for (std::vector<region_info>::const_iterator iter =
+                 regions.begin(); iter != regions.end(); ++iter) {
+            size_t size = UCS_PTR_BYTE_DIFF(iter->start, iter->end);
+            test_ptr_released(iter->start, size);
+            test_ptr_not_found(iter->start, size);
```

Comment:
does it call ucs_pgtable_search_range()?

Reply:
yes, it's done implicitly inside `memtype_cache_remove()` and `memtype_cache_update()`

Reply:
i see it only calls ucs_memtype_cache_lookup() which calls only ucs_pgtable_lookup(). what am i missing?

Reply:
We call `ucs_pgtable_search_range()` through the foloowing flows:
`memtype_cache_update()` -> `ucs_memtype_cache_update()` -> `ucs_memtype_cache_update_internal(UCS_MEMTYPE_CACHE_ACTION_SET_MEMTYPE)` -> `ucs_pgtable_search_range()`
`memtype_cache_remove()` -> `ucs_memtype_cache_remove()` -> `ucs_memtype_cache_update_internal(UCS_MEMTYPE_CACHE_ACTION_REMOVE)` -> `ucs_pgtable_search_range()`
in order to add/remove regions.

We call `ucs_memtype_cache_lookup()` from tests in order to check whether some region in Cache or not

Reply:
right.